### ISSUE TITLE
RNNT loss Umax mismatch error

### ISSUE THREAD

**Issue by OllieBroadhurst** [NVIDIA/NeMo#6040]
**Describe the bug**

Hi! I'm having an issue while fine-tuning `stt_en_conformer_transducer_large`. After a while I get an error stating:
`Output length mismatch! Given U: 35, Expected max U from target lengths: 35 + 1` triggered [here](https://github.com/NVIDIA/NeMo/blob/3e7499599a13869e772612d286d660b574715a5e/nemo/collections/asr/parts/numba/rnnt_loss/rnnt_pytorch.py#L405).

In my mind this is strange because I was already 4 epochs in and I'd imagine the loss has already been calculated for all samples. My understanding of the RNNT loss itself is limited though so I might be missing something.

The maximum number of tokens in the dataset labels is 39 (using the sp tokenizer). Training was not resumed from a previous checkpoint. The architecture of the model has stayed the same during training.

**Steps/Code to reproduce bug**

Normal training with own dataset and custom tokenizer. Max token length in the dataset: 39 (a clue?).

Train config:
```yaml
  loss:
    loss_name: default
    warprnnt_numba_kwargs:
      fastemit_lambda: 0.0
      clamp: -1.0
  variational_noise:
    start_step: 0
    std: 0.0
  optim:
    name: adamw
    lr: 5.0
    betas:
    - 0.9
    - 0.98
    weight_decay: 0.001
    sched:
      name: NoamAnnealing
      d_model: 512
      warmup_steps: 1000
      warmup_ratio: null
      min_lr: 1.0e-06
```

**Expected behavior**

Train loss should be calculated normally.

**Environment overview (please complete the following information)**

 - Environment location: GCP VM
 - Method of NeMo install: `pip install nemo-toolkit`

**Environment details**

If NVIDIA docker image is used you don't need to specify these.
Otherwise, please provide:
- OS: Ubuntu 20.04
- PyTorch version: 1.12.1+cu113
- Python version: 3.7.12

**Additional context**

GPU: 1xV100

**Comment by titu1994** [comment]
Hmm that is weird. 

That assertion is for the Blank token used as Beginning of Sentence - whatever length of text (Char or Subword), it gets prepended by a blank token ID for BOS. Basically your joint tensor has the shape (B,T,U,V+1) and your max label seq length is U but the expected shape is (B,T,U+1,V+1) when given max label seq length U. Either somehow your label seq length is off by 1 or your longest text token is off by one somehow. Both cases should not be possible in regular usage unless there's some odd character that doesn't get counted when you tokenize the text and then call len(list of subwords) on it. 

It's also odd that it takes till 4th epochs to hit this issue.

I don't really have a solution for this since I've never encountered this case where text seq length isn't correct.

**Comment by OllieBroadhurst** [comment]
Thanks for the reply and explanation!

As it turns out I may have found the issue. My tokenizer `spe_type` was `unigram` when the model typically expects `bpe`. So far I'm in epoch 5 and nothing nefarious has happened.

The number of tokens can vary a lot depending on the subword type, could this be unexpected by the model? I'm not sure how it's doing this from a loss calculation point of view as `target_length` should adjust to the encoded length regardless of the method used.

I can't explain the random nature of the problem either. If there's shuffling between batches, maybe there's a combination of mis-encoded targets every now and then which could affect the target lengths.

I'll keep this open for a little while until I'm 100% happy this was the solution.

**Comment by titu1994** [comment]
Tokenizer type of unigram usually is sufficient for most datasets with sufficient text. I suppose the Tokenizer type has a minor effect on wer but it usually does not impact loss calculation. 

Perhaps if you dataset is small enough, that vocab size can't actually match your tokenizer size (for example, unigram tries to build 1024 tokens but finds there are only 1000 unique unigrams, then that can cause an error. But I doubt that's what happened here.

**Comment by titu1994** [comment]
Model type is just a broad term for algorithm. It is an umbrella term for spe bpe, spe unigram, spe char and spe word all together under "BPE" since that's the underlying algorithm. Perhaps better to call it "spe" than "bpe" in the tokenizer config. I will add a note to support that.

**Comment by OllieBroadhurst** [comment]
Just demonstrates how many decisions need to be taken into account before training begins. Thanks for your help. I'm happy this was the solution unless I get really unlucky.

Great model, great repo.
